python/ray/data/_internal/pandas_block.py

```diff
@@ -98,31 +96,18 @@ def _table_from_pydict(self, columns: Dict[str, List[Any]]) -> "pandas.DataFrame
 
     def _concat_tables(self, tables: List["pandas.DataFrame"]) -> "pandas.DataFrame":
         pandas = lazy_import_pandas()
-        from ray.data.extensions.tensor_extension import TensorArray
+        from ray.air.util.data_batch_conversion import (
+            _cast_ndarray_columns_to_tensor_extension,
+        )
 
         if len(tables) > 1:
             df = pandas.concat(tables, ignore_index=True)
         else:
             df = tables[0]
-        # Try to convert any ndarray columns to TensorArray columns.
-        # TODO(Clark): Once Pandas supports registering extension types for type
-        # inference on construction, implement as much for NumPy ndarrays and remove
-        # this. See https://github.com/pandas-dev/pandas/issues/41848
-        for col_name, col in df.items():
-            if (
-                col.dtype.type is np.object_
-                and not col.empty
-                and isinstance(col.iloc[0], np.ndarray)
-            ):
-                try:
-                    df.loc[:, col_name] = TensorArray(col)
-                except Exception as e:
-                    if ray.util.log_once("datasets_tensor_array_cast_warning"):
-                        logger.warning(
-                            f"Tried to transparently convert column {col_name} to a "
-                            "TensorArray but the conversion failed, leaving column "
-                            f"as-is: {e}"
-                        )
+        df.reset_index(drop=True, inplace=True)
```

Comment:
Slicing opaque object columns fails silently in user-code if we don't reset the index here (Pandas foot-gun that our tensor extension type avoids).